File: src/Day02.java
import java.nio.file.Path;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.stream.Gatherers;
import java.util.stream.Stream;

import static java.lang.Integer.signum;
import static java.lang.Math.abs;
import static java.nio.file.Files.readAllLines;
import static java.util.stream.IntStream.range;
import static java.util.stream.Stream.concat;

public class Day02 {
    public static void main(String[] args) throws Exception {
        List<String> input;
        if (args.length == 0) {
            assert part1(readAllLines(Path.of("src/2t.txt"))) == 2;
            input = readAllLines(Path.of("src/2.txt"));
        } else input = readAllLines(Path.of(args[0]));
        System.out.println(part1(input));
        System.out.println(part2(input));
    }

    private static Stream<Integer[]> parse(List<String> input) {
        return input
            .stream()
            .map(line -> Stream
                .of(line.split(" "))
                .map(Integer::parseInt)
                .toArray(Integer[]::new));
    }

    @SuppressWarnings("preview")
    private static boolean validReport(Integer[] report) {
        int dir = signum(report[1] - report[0]);
        if (dir == 0) return false;
        return Stream
            .of(report)
            .gather(Gatherers.windowSliding(2))
            .noneMatch(l -> {
                int diff = l.get(1) - l.get(0);
                return abs(diff) > 3 || signum(diff) != dir;
            });
    }

    private static int part1(List<String> input) {
        return (int) parse(input).filter(Day02::validReport).count();
    }

    private static int part2(List<String> input) {
        return (int)
            parse(input)
                .filter(report -> concat(
                    Stream.ofNullable(report),
                    range(0, report.length)
                        .mapToObj(index -> withIndexRemoved(report, index))
                )
                    .anyMatch(Day02::validReport))
                .count();
    }

    private static Integer[] withIndexRemoved(Integer[] report, int index) {
        var modifiedReport = new ArrayList<>(Arrays.asList(report));
        modifiedReport.remove(index);
        return modifiedReport.toArray(Integer[]::new);
    }

}

File: src/Day02.kt
import kotlin.io.path.Path
import kotlin.io.path.readText
import kotlin.math.abs
import kotlin.math.sign

fun main(args: Array<String>) {
    fun parseDay1(input: List<String>) = input
        .map { it.split(" ").map(String::toInt) }

    fun validReport(report: List<Int>): Boolean {
        val dir = (report[1] - report[0]).sign
        if (dir == 0) return false
        for (i in 1 until report.size) {
            val diff = report[i] - report[i - 1]
            if (abs(diff) > 3 || diff.sign != dir.sign)
                return false
        }
        return true
    }

    fun part1(input: List<String>): Int {
        val reports = parseDay1(input)
        return reports.count(::validReport)
    }

    fun part2(input: List<String>): Int {
        val reports = sequence {
            val data = parseDay1(input)
            for (report in data) {
                yield(sequence {
                    yield(report)
                    yieldAll(report.indices.asSequence().map { index ->
                        report.toMutableList().also { it.removeAt(index) }.toList()
                    })
                })
            }
        }

        return reports.count {
            it.any(::validReport)
        }
    }

    // Test if implementation meets criteria from the description, like:
//    check(part1(listOf("test_input")) == 1)

    // Or read a large test input from the `src/Day01_test.txt` file:
    if (args.isEmpty()) {
        val testInput = readInput("2t")
        check(part1(testInput) == 2)

        // Read the input from the `src/Day01.txt` file.
        val input = readInput("2")
        part1(input).println()
        part2(input).println()
    } else {
        val input = Path(args[0]).readText().trim().lines()
        part1(input).println()
        part2(input).println()
    }
}


